### ISSUE TITLE
HOL generation failure

### ISSUE THREAD

**Issue by dpgao** [CTSRD-CHERI/sail-cheri-riscv#44]
I'm on `sail -v` = `Sail 0.13 (sail2 @ opam)`.

Running `make riscv_hol_build` results in the following error:

```
mkdir -p generated_definitions/lem/RV64 generated_definitions/isabelle/RV64
sail  -lem -lem_output_dir generated_definitions/lem/RV64 -isa_output_dir generated_definitions/isabelle/RV64 -o riscv -lem_mwords -lem_lib Riscv_extras -lem_lib Riscv_extras_fdext -lem_lib Cheri_extras -no_effects -mono_rewrites /Users/dpgao/.opam/default/share/sail/lib/mono_rewrites.sail sail-riscv/model/prelude.sail sail-riscv/model/prelude_mapping.sail sail-riscv/model/riscv_xlen64.sail sail-riscv/model/riscv_flen_D.sail src/cheri_prelude.sail src/cheri_types.sail src/cheri_prelude_128.sail src/cheri_mem_metadata.sail sail-riscv/model/prelude_mem.sail src/cheri_cap_common.sail sail-riscv/model/riscv_types_common.sail src/cheri_riscv_types.sail sail-riscv/model/riscv_types.sail src/cheri_reg_type.sail sail-riscv/model/riscv_freg_type.sail sail-riscv/model/riscv_csr_map.sail src/cheri_scr_map.sail src/cheri_vmem_types.sail sail-riscv/model/riscv_regs.sail sail-riscv/model/riscv_sys_regs.sail sail-riscv/model/riscv_pmp_regs.sail sail-riscv/model/riscv_pmp_control.sail src/cheri_sys_regs.sail src/cheri_regs.sail src/cheri_pc_access.sail sail-riscv/model/riscv_next_regs.sail src/cheri_sys_exceptions.sail sail-riscv/model/riscv_sync_exception.sail sail-riscv/model/riscv_next_control.sail sail-riscv/model/riscv_softfloat_interface.sail sail-riscv/model/riscv_fdext_regs.sail sail-riscv/model/riscv_fdext_control.sail sail-riscv/model/riscv_csr_ext.sail sail-riscv/model/riscv_sys_control.sail sail-riscv/model/riscv_addr_checks_common.sail src/cheri_addr_checks.sail src/cheri_misa_ext.sail sail-riscv/model/riscv_platform.sail sail-riscv/model/riscv_mem.sail src/cheri_mem.sail src/cheri_pte.sail src/cheri_ptw.sail sail-riscv/model/riscv_vmem_common.sail sail-riscv/model/riscv_vmem_tlb.sail sail-riscv/model/riscv_vmem_sv39.sail sail-riscv/model/riscv_vmem_sv48.sail sail-riscv/model/riscv_vmem_rv64.sail sail-riscv/model/riscv_insts_begin.sail sail-riscv/model/riscv_insts_base.sail sail-riscv/model/riscv_insts_aext.sail sail-riscv/model/riscv_insts_cext.sail sail-riscv/model/riscv_insts_mext.sail sail-riscv/model/riscv_insts_zicsr.sail sail-riscv/model/riscv_insts_fext.sail sail-riscv/model/riscv_insts_dext.sail src/cheri_insts.sail sail-riscv/model/riscv_jalr_seq.sail  sail-riscv/model/riscv_insts_end.sail  sail-riscv/model/riscv_step_common.sail src/cheri_step_ext.sail src/cheri_decode_ext.sail sail-riscv/model/riscv_fetch.sail sail-riscv/model/riscv_step.sail sail-riscv/model/riscv_analysis.sail -splice handwritten_support/dummy_assembly_mappings.sail
Warning: Possible incomplete pattern match at
Most general matched pattern is AT_CAP_ERR_

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:76:7-12
76 |  then match n {
   |       ^---^
   | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:82:7-12
82 |  else match n { //  sizeof(xlen) == 64
   |       ^---^
   | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:102:7-12
102 |  then match n {
    |       ^---^
    | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:125:7-12
125 |  then match n {
    |       ^---^
    | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at
Most general matched pattern is AT_CAP_ERR_

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:76:7-12
76 |  then match n {
   |       ^---^
   | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:82:7-12
82 |  else match n { //  sizeof(xlen) == 64
   |       ^---^
   | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:102:7-12
102 |  then match n {
    |       ^---^
    | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:125:7-12
125 |  then match n {
    |       ^---^
    | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at
Most general matched pattern is AT_CAP_ERR_

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:76:7-12
76 |  then match n {
   |       ^---^
   | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:82:7-12
82 |  else match n { //  sizeof(xlen) == 64
   |       ^---^
   | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:102:7-12
102 |  then match n {
    |       ^---^
    | 
Most general matched pattern is 0

Warning: Possible incomplete pattern match at [sail-riscv/model/riscv_pmp_regs.sail]:125:7-12
125 |  then match n {
    |       ^---^
    | 
Most general matched pattern is 0

echo "declare {isabelle} rename field sync_exception_ext = sync_exception_ext_exception" >> generated_definitions/lem/RV64/riscv_types.lem
lem -hol -outdir generated_definitions/hol4/RV64 -lib /Users/dpgao/.opam/default/share/sail/lib/hol -i /Users/dpgao/.opam/default/share/sail/lib/hol/sail2_prompt_monad.lem -i /Users/dpgao/.opam/default/share/sail/lib/hol/sail2_prompt.lem \
	    -lib /Users/dpgao/.opam/default/share/sail/src/lem_interp -lib /Users/dpgao/.opam/default/share/sail/src/gen_lib \
		sail-riscv/handwritten_support/0.11/riscv_extras.lem sail-riscv/handwritten_support/0.11/riscv_extras_fdext.lem \
		handwritten_support/cheri_extras.lem \
		generated_definitions/lem/RV64/riscv_types.lem \
		generated_definitions/lem/RV64/riscv.lem
File "generated_definitions/lem/RV64/riscv_types.lem", line 207, character 22 to line 207, character 31
  Type error: unbound type constructor: AT_CAP_ERR
make: *** [generated_definitions/hol4/RV64/riscvScript.sml] Error 1
```

**Comment by bacam** [comment]
This is a Sail bug, which I've fixed in rems-project/sail@3d0064a157a711f21d23f56dc2c423b2e3b1d74a.  By the way, if you have trouble building the resulting HOL script, try adding the `-grouped_regstate` option to the sail command (which we should probably make the default).

**Comment by dpgao** [comment]
@bacam Thanks for the fix! There are no more sail errors now but the following error appeared at the `Holmake` stage, both with and without `-grouped_regstate`. I'm on commit a1da534 of the spec.

```
riscvTheory                                                                                                       FAILED! <1>
 
 can not be constrained to be of type
 
 :ext_ptw
 
 unification failure message: ???
 
 error in quse /Users/dpgao/src/sail-cheri-riscv/generated_definitions/hol4/RV64/riscvScript.sml : HOL_ERR {message = "on line 18644, characters 39-45:\n\nType inference failure: the term\n\next_ptw :ext_ptw_lc -> ext_ptw_sc -> ext_ptw\n\non line 18644, characters 39-45\n\ncan not be constrained to be of type\n\n:ext_ptw\n\nunification failure message: ???\n", origin_function = "type-analysis", origin_structure = "Preterm"}
 error in load riscvScript : HOL_ERR {message = "on line 18644, characters 39-45:\n\nType inference failure: the term\n\next_ptw :ext_ptw_lc -> ext_ptw_sc -> ext_ptw\n\non line 18644, characters 39-45\n\ncan not be constrained to be of type\n\n:ext_ptw\n\nunification failure message: ???\n", origin_function = "type-analysis", origin_structure = "Preterm"}
 Uncaught exception: HOL_ERR {message = "on line 18644, characters 39-45:\n\nType inference failure: the term\n\next_ptw :ext_ptw_lc -> ext_ptw_sc -> ext_ptw\n\non line 18644, characters 39-45\n\ncan not be constrained to be of type\n\n:ext_ptw\n\nunification failure message: ???\n", origin_function = "type-analysis", origin_structure = "Preterm"}
```

**Comment by bacam** [comment]
The problem is that `ext_ptw` is used as both the name of a record (and hence the record's constructor) and a variable.  HOL doesn't like this, so we ought to rename it in Lem.  I won't be able to do this quickly, however, so if you need to make some progress you could rename the variable yourself, either in the Sail source file `src/cheri_pte.sail` or the generated HOL file.

**Comment by dpgao** [comment]
Thanks @bacam. I had to rename occurrances of `ext_ptw` not only in `src/cheri_pte.sail` but also in `sail-riscv/model/riscv_vmem_*.sail` in the upstream spec to successfully complete the build.

**Comment by bacam** [comment]
I've fixed this in rems-project/lem@204702c and checked that it works with `make riscv_hol_build SAIL_FLAGS+=-grouped_regstate`.
